### PR TITLE
Fix: DNA Infuser & Unit Tests, Organs Bugfixes

### LINKED ISSUE #72767: Gene Infuser doesn't splice genes
Reporting client version: 514.1589

## Round ID:
198368

## Testmerges:
- [Makes nerve splicing surgery reduce stamina damage](https://www.github.com/tgstation/tgstation/pull/72661)

## Reproduction:
Try to splice yourself with DNA of a dead rat or goliath, nothing happens

### PR DESCRIPTION
**Pull request by Floofies**
## About The Pull Request

>_"I don't remember buying tickets to Mutants on Ice."_
>-Duke Nukem

This PR is (hopefully the final) part of a series of my continuing refactors of the DNA Infuser. This PR represents a "quality pass" which should also iron-out the rest of the most impactful bugs.

Granular list of changes:
- This PR adds unit tests for the DNA Infuser organs and `/datum/status_effect/organ_set_bonus` as recommended by @AnturK
- I noticed that the base `/datum/infuser_entry` was being used in the machine for the Fly and "rejected" infusions, whereas usually we would expect it to be a base type used only as a development template. I corrected this issue and created `/datum/infuser_entry/fly` to be used for that use-case instead.
- Added `/mob/proc/can_mutate()` and `/mob/living/carbon/can_mutate()` to replace a few copied lines across several files. The proc is normally used in the context of mutating a Human via their DNA.
- I fixed a ton of typos in organ-related code, specifically where "receiver" was typo'd as "reciever". There are far more of those typos, but I limited the scope of my changes to organs.
- I noticed a bug in `/datum/species/proc/regenerate_organs`  wherein a race condition caused an organ to remove itself before it's done inserting itself. This happens because the Fly organ set bonus runs `regenerate_organs` which calls `Remove` on the organ while `Insert` is still in the call-stack. I added `INVOKE_ASYNC` as a workaround, and also changed the order the signals are emitted to prevent future bugs. This bug primarily only impacted the flyperson species transformation, which was part of the DNA Infuser's flyperson infusion organ set bonus.
- In my last refactor PR #72745 I also introduced a bug in `/obj/machinery/dna_infuser/proc/infuse_organ` wherein I forgot to add the usage of `new` when attempting to implant new organs, and this PR fixes the erroneous code.
- Fxed a bug which causes the organ set bonus to activate when mixing organs from different sources, which is caused by a developer oversight wherein all `/datum/status_effect/organ_set_bonus`  had identical IDs.
- Added a cleaner `replacetext`-based way of handling pronouns in `/datum/element/noticable_organ/proc/on_receiver_examine`, using custom macros `%PRONOUN_S` and `%PRONOUN_ES` as advised by @MrMelbert 
- This PR also fixes #72767

## Why It's Good For The Game

With the changes in this PR the machine will finally work as we expect it to. By adding unit tests we will also be able to ensure that it works as expected from now on. I feel confident saying that the completeness, algorithmic correctness, and code health of the DNA Infuser is much better than it was before.

## Changelog

:cl: A.C.M.O.
fix: Fully fixed the DNA Infuser, which will now infuse organs as expected.
fix: Fixed flyperson species transformation and organ set bonus, which was throwing a runtime.
fix: Fixed many typos in organ-related source code.
/:cl:

### CODE REVIEW

**Review by tralezab — CHANGES_REQUESTED** [review]

File: code/datums/elements/organ_set_bonus.dm [hunk 1]
```diff
@@ -6,17 +6,14 @@
...
-/datum/element/organ_set_bonus/Attach(datum/target, bonus_type)
+/datum/element/organ_set_bonus/Attach(datum/target, bonus_status)
 	. = ..()
 
 	if(!isorgan(target))
 		return ELEMENT_INCOMPATIBLE
-
-	src.bonus_type = bonus_type
-
+	bonus_type = bonus_status
```
[the remark this answers is not in the document]
Comment (Floofies) [reply code/datums/elements/organ_set_bonus.dm]: Fixed by commit 6bc7593ed49539bea89ee97440cf520f18958f69

**Comment by Floofies** [comment]
Fixing my unit tests now... looks like a race condition wherein an organ tries to remove itself before it's even done inserting itself. This happens because the Fly organ set bonus runs `regenerate_organs()` which calls `Remove()` on the organ while `Insert()` is still in the call-stack

**Review by Time-Green — CHANGES_REQUESTED** [review]
Mostly some cosmetic stuff, nice work!

File: code/game/machinery/dna_infuser/dna_infuser.dm [hunk 2]
```diff
@@ -98,33 +98,38 @@
...
-		// Organ IS brain, insert via special logic:
-		var/obj/item/organ/internal/brain/old_brain = target.getorganslot(ORGAN_SLOT_BRAIN)
-		// Brains REALLY like ghosting people. we need special tricks to avoid that, namely removing the old brain with no_id_transfer
-		old_brain.Remove(target, special = TRUE, no_id_transfer = TRUE)
-		qdel(old_brain)
-		var/obj/item/organ/internal/brain/new_brain = new_organ
-		new_brain.Insert(target, special = TRUE, drop_if_replaced = FALSE, no_id_transfer = TRUE)
+		return TRUE
+	// Organ IS brain, insert via special logic:
+	var/obj/item/organ/internal/brain/old_brain = target.getorganslot(ORGAN_SLOT_BRAIN)
```
[the remark this answers is not in the document]
Comment (Floofies) [reply code/game/machinery/dna_infuser/dna_infuser.dm:113]: The machine will only replace a pre-existing organ. Player's should perceive that the machine is "mutating" the mob's organs and not giving them entirely new organs.

[the remark this answers is not in the document]
Comment (Time-Green) [reply code/game/machinery/dna_infuser/dna_infuser.dm:113]: Ah okay

File: code/game/machinery/dna_infuser/infuser_entries.dm [hunk 3]
```diff
@@ -1,16 +1,16 @@
 /// A list of all infuser entries
-GLOBAL_LIST_INIT(infuser_entries, prepare_entries())
+GLOBAL_LIST_INIT(infuser_entries, prepare_infuser_entries())
 
 /// just clarifying that no threshold does some special stuff, since only meme mutants have it
-#define NO_THRESHOLD ""
+#define DNA_INFUSION_NO_THRESHOLD ""
 
-/proc/prepare_entries()
+/proc/prepare_infuser_entries()
```
[the remark this answers is not in the document]
Comment (Floofies) [reply code/game/machinery/dna_infuser/infuser_entries.dm:8]: Fixed by commit c886fc2bc31885af85795a4729a33a6a3a273db2

File: code/game/machinery/dna_infuser/organ_sets/carp_organs.dm [hunk 4]
```diff
@@ -61,24 +52,26 @@
 	if(!ishuman(tongue_owner))
 		return
 	var/mob/living/carbon/human/human_receiver = tongue_owner
-	var/datum/species/rec_species = human_receiver.dna.species
-	rec_species.update_no_equip_flags(tongue_owner, rec_species.no_equip_flags | ITEM_SLOT_MASK)
-	var/obj/item/bodypart/head/head = human_receiver.get_bodypart(BODY_ZONE_HEAD)
-	head.unarmed_damage_low = 10
-	head.unarmed_damage_high = 15
-	head.unarmed_stun_threshold = 15
+	if(human_receiver.can_mutate())
```
[the remark this answers is not in the document]
Comment (Floofies) [reply code/game/machinery/dna_infuser/organ_sets/carp_organs.dm]: Fixed by commit e792ab8df2e4bab4d4d826fa3f54d137ba4d545b

File: code/game/machinery/dna_infuser/organ_sets/carp_organs.dm [hunk 5]
```diff
@@ -61,24 +52,26 @@
...
 	if(!ishuman(tongue_owner))
 		return
 	var/mob/living/carbon/human/human_receiver = tongue_owner
-	var/datum/species/rec_species = human_receiver.dna.species
-	rec_species.update_no_equip_flags(tongue_owner, initial(rec_species.no_equip_flags))
-	var/obj/item/bodypart/head/head = human_receiver.get_bodypart(BODY_ZONE_HEAD)
-	head.unarmed_damage_low = initial(head.unarmed_damage_low)
-	head.unarmed_damage_high = initial(head.unarmed_damage_high)
-	head.unarmed_stun_threshold = initial(head.unarmed_stun_threshold)
+	if(human_receiver.can_mutate())
```
[the remark this answers is not in the document]
Comment (Floofies) [reply code/game/machinery/dna_infuser/organ_sets/carp_organs.dm]: Fixed by commit e792ab8df2e4bab4d4d826fa3f54d137ba4d545b

File: code/game/machinery/dna_infuser/organ_sets/goliath_organs.dm [hunk 6]
```diff
@@ -82,24 +68,24 @@
 	if(!ishuman(brain_owner))
 		return
 	var/mob/living/carbon/human/human_receiver = brain_owner
-	var/datum/species/rec_species = human_receiver.dna.species
-	rec_species.update_no_equip_flags(brain_owner, rec_species.no_equip_flags | ITEM_SLOT_GLOVES)
-
-	hammer = new/obj/item/goliath_infuser_hammer
-	brain_owner.put_in_hands(hammer)
+	if(human_receiver.can_mutate())
```
[the remark this answers is not in the document]
Comment (Floofies) [reply code/game/machinery/dna_infuser/organ_sets/goliath_organs.dm]: Fixed by commit e792ab8df2e4bab4d4d826fa3f54d137ba4d545b

File: code/game/machinery/dna_infuser/organ_sets/rat_organs.dm [hunk 7]
```diff
@@ -55,23 +43,22 @@
...
-	var/mob/living/carbon/human/human_holder = reciever
-	var/datum/species/species = human_holder.dna.species
-	//mmm, cheese. doesn't especially like anything else
-	species.liked_food = DAIRY
-	//but a rat can eat anything without issue
-	species.disliked_food = NONE
-	species.toxic_food = NONE
-	human_holder.physiology.hunger_mod *= 10
+	var/mob/living/carbon/human/human_holder = receiver
+	if(human_holder.can_mutate())
```
[the remark this answers is not in the document]
Comment (Floofies) [reply code/game/machinery/dna_infuser/organ_sets/rat_organs.dm]: Fixed by commit e792ab8df2e4bab4d4d826fa3f54d137ba4d545b

File: code/game/machinery/dna_infuser/organ_sets/rat_organs.dm [hunk 8]
```diff
@@ -85,15 +72,15 @@
 	if(!ishuman(stomach_owner))
 		return
 	var/mob/living/carbon/human/human_holder = stomach_owner
-	var/datum/species/species = human_holder.dna.species
-	species.liked_food = initial(species.liked_food)
-	species.disliked_food = initial(species.disliked_food)
-	species.toxic_food = initial(species.toxic_food)
-	human_holder.physiology.hunger_mod /= 10
+	if(human_holder.can_mutate())
```
[the remark this answers is not in the document]
Comment (Floofies) [reply code/game/machinery/dna_infuser/organ_sets/rat_organs.dm]: Fixed by commit e792ab8df2e4bab4d4d826fa3f54d137ba4d545b

File: code/game/machinery/dna_infuser/organ_sets/rat_organs.dm [hunk 9]
```diff
@@ -85,15 +72,15 @@
...
+		var/datum/species/species = human_holder.dna.species
+		species.liked_food = initial(species.liked_food)
+		species.disliked_food = initial(species.disliked_food)
+		species.toxic_food = initial(species.toxic_food)
+		if(human_holder.physiology)
+			human_holder.physiology.hunger_mod /= 10
```
[the remark this answers is not in the document]
Comment (Floofies) [reply code/game/machinery/dna_infuser/organ_sets/rat_organs.dm]: Fixed by commit 77fbd401843ae924aa973d7a99c02dd62dba734a

**Comment by Floofies** [comment]
@Time-Green Ok, this is ready to merge!

**Pull request merged**